internal/events/events.go

```diff
@@ -2,115 +2,121 @@ package events
+	}
 	// Convert the hex-encoded data to bytes
 	trimmed := strings.TrimPrefix(log.Data, "0x")
 	data, err := hex.DecodeString(trimmed)
 
 	if err != nil {
-		return nil, err
+		return "", nil, err
 	}
 
-	contractABI, error := abi.JSON(strings.NewReader(dpos_interface.DposInterfaceABI))
+	contractABI, error := abi.JSON(strings.NewReader(relevantAbi))
 	if error != nil {
-		return nil, error
+		return "", nil, error
 	}
 
-	rewardsClaimedTopic := crypto.Keccak256Hash([]byte("RewardsClaimed(address,address,uint256)"))
-	commissionRewardsClaimedTopic := crypto.Keccak256Hash([]byte("CommissionRewardsClaimed(address,address,uint256)"))
-	// Decode the event based on its topic
-	switch log.Topics[0] {
-	case rewardsClaimedTopic.Hex():
-		var event RewardsClaimedEvent
-		err := contractABI.UnpackIntoInterface(&event, "RewardsClaimed", data)
+	// Get the event for the topic
 
-		if err != nil {
-			return nil, err
-		}
-		account := ethcommon.HexToAddress(log.Topics[1])
-		validator := ethcommon.HexToAddress(log.Topics[2])
+	event, err := contractABI.EventByID(ethcommon.HexToHash(log.Topics[0]))
 
-		// Set the addresses in the event struct
-		event.Account = account.Hex()
-		event.Validator = validator.Hex()
-		return &event, nil
+	if err != nil {
+		return "", nil, err
+	}
 
-	case commissionRewardsClaimedTopic.Hex():
-		var event CommissionRewardsClaimedEvent
-		err := contractABI.UnpackIntoInterface(&event, "CommissionRewardsClaimed", data)
-		if err != nil {
-			return nil, err
-		}
-		account := ethcommon.HexToAddress(log.Topics[1])
-		validator := ethcommon.HexToAddress(log.Topics[2])
+	unpacked, err := contractABI.Unpack(event.Name, data)
 
-		// Set the addresses in the event struct
-		event.Account = account.Hex()
-		event.Validator = validator.Hex()
-		return &event, nil
+	if err != nil {
+		return "", nil, err
+	}
+
+	params, err := parseToStringSlice(unpacked)
+
+	if err != nil {
+		return "", nil, err
 	}
-	return nil, nil
+
+	return event.Sig, params, nil
 }
 
 func DecodeRewardsTopics(logs []models.EventLog) (decodedEvents []LogReward, err error) {
 	for _, log := range logs {
 		if !strings.EqualFold(log.Address, common.DposContractAddress) {
 			continue
 		}
-		decoded, err := DecodeEvent(log)
+		name, decoded, err := DecodeEventDynamic(log)
 		if err != nil {
 			return nil, err
 		}
 
-		switch event := decoded.(type) {
-		case *RewardsClaimedEvent:
-			decodedEvents = append(decodedEvents, LogReward{
-				EventName: "RewardsClaimed",
-				Account:   event.Account,
-				Validator: event.Validator,
-				Value:     event.Amount,
-			})
-
-		case *CommissionRewardsClaimedEvent:
-			decodedEvents = append(decodedEvents, LogReward{
-				EventName: "CommissionRewardsClaimed",
-				Account:   event.Account,
-				Validator: event.Validator,
-				Value:     event.Amount,
-			})
+		fmt.Println(name)
+
+		if name == rewardsClaimedName || name == commissionRewardsClaimedName {
+			account := ethcommon.HexToAddress(log.Topics[1])
```

Comment:
If I am not mistaken this addresses should be already parsed and sliced in the `decoded`.

Reply:
Nope, as they are not part of the data. Indexed addresses are added to the end of the topics